File: src/main/scala/gameover/fwk/ai/CollisionDetector.scala
package gameover.fwk.ai

import com.badlogic.gdx.math.{Rectangle, Vector2}
import com.badlogic.gdx.utils.{Array => GdxArray}

trait CollisionDetector {

  /**
   * Check an area status
   */
  def checkPosition(area: Rectangle): CollisionState.Value

  /**
   * Check a position status
   */
  def checkPosition(x: Float, y: Float) : CollisionState.Value

  /**
   * This method check if area is intersecting something it shouldn't and return an array of
   * Rectangle objects where the collision append. The velocity of the sprite is updated accordingly.
   */
  def checkCollision(area: Rectangle, movingSpriteVelocity: Vector2, onlyBlocking: Boolean): GdxArray[Rectangle]

  /**
    * Check if a point has a direct view to a target point.
    */
  def isDirect(x: Float, y: Float, targetX: Float, targetY: Float, onlyBlocking: Boolean): Boolean

  /**
   * Check if a sprite defined by its area has a direct view to a target point.
   */
  def isDirect(visionArea: Rectangle, targetX: Float, targetY: Float, onlyBlocking: Boolean): Boolean

  /**
    * Check if a sprite defined by its area has a direct view to a target area.
    */
  def isDirect(visionArea: Rectangle, targetArea: Rectangle, onlyBlocking: Boolean): Boolean

  def checkCollision(area: Rectangle, onlyBlocking: Boolean) : Boolean = {
    val state: CollisionState.Value = checkPosition(area)
    (onlyBlocking && state == CollisionState.Blocking) || (!onlyBlocking && state != CollisionState.Empty)
  }

  def checkCollision(x: Float, y: Float, onlyBlocking: Boolean) : Boolean = {
    val state: CollisionState.Value = checkPosition(x, y)
    (onlyBlocking && state == CollisionState.Blocking) || (!onlyBlocking && state != CollisionState.Empty)
  }
}

object CollisionState extends Enumeration {
  val Blocking, Void, Empty = Value
}

File: src/main/kotlin/gameover/fwk/ai/CollisionDetector.kt
package gameover.fwk.ai

import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2

interface CollisionDetector {

    /**
     * Check an area status
     */
    fun checkPosition(area: Rectangle): CollisionState

    /**
     * Check a position status
     */
    fun checkPosition(x: Float, y: Float) : CollisionState

    /**
     * This method check if area is intersecting something it shouldn't and return an array of
     * Rectangle objects where the collision append. The velocity of the sprite is updated accordingly.
     */
    fun checkCollision(area: Rectangle, movingSpriteVelocity: Vector2, onlyBlocking: Boolean): List<Rectangle>

    /**
     * Check if a point has a direct view to a target point. Only blocking can block this view.
     */
    fun hasDirectView(x: Float, y: Float, targetX: Float, targetY: Float): Boolean

    /**
     * Check if a sprite defined by its area has a direct view to a target point. Only blocking can block this view.
     */
    fun hasDirectView(visionArea: Rectangle, targetX: Float, targetY: Float): Boolean

    /**
     * Check if a sprite defined by its area has a direct view to a target area. Only blocking can block this view.
     */
    fun hasDirectView(visionArea: Rectangle, targetArea: Rectangle): Boolean

    /**
     * Check if a sprite can move straight to a point.
     */
    fun canMoveStraightToPoint(area: Rectangle, targetX: Float, targetY: Float): Boolean

    /**
     * Check area vs the collision map to see if something collides.
     */
    fun checkCollision(area: Rectangle, onlyBlocking: Boolean) : Boolean {
        val state: CollisionState = checkPosition(area)
        return (onlyBlocking && state == CollisionState.Blocking) || (!onlyBlocking && state != CollisionState.Empty)
    }

    /**
     * Check a point vs the collision map to see if something collides.
     */
    fun checkCollision(x: Float, y: Float, onlyBlocking: Boolean) : Boolean {
        val state: CollisionState = checkPosition(x, y)
        return (onlyBlocking && state == CollisionState.Blocking) || (!onlyBlocking && state != CollisionState.Empty)
    }
}

enum class CollisionState { Blocking, Void, Empty }
